### PR TITLE
Multiple permission roots

### PR DESCRIPTION
**Pull request by SofiaSazonova**
### Feature or Bugfix
- Bugfix

### Detail
- GET_DATASET_TABLE (FOLDER) permissions are granted to the group only if they are not granted already
- these permissions are removed if group is not admin|steward and there are no other shares of this item.

### Relates
- #1174

### Security
Please answer the questions below briefly where applicable, or write `N/A`. Based on
[OWASP 10](https://owasp.org/Top10/en/).

- Does this PR introduce or modify any input fields or queries - this includes
fetching data from storage outside the application (e.g. a database, an S3 bucket)?
  - Is the input sanitized?
  - What precautions are you taking before deserializing the data you consume?
  - Is injection prevented by parametrizing queries?
  - Have you ensured no `eval` or similar functions are used?
- Does this PR introduce any functionality or component that requires authorization?
  - How have you ensured it respects the existing AuthN/AuthZ mechanisms?
  - Are you logging failed auth attempts?
- Are you using or adding any cryptographic features?
  - Do you use a standard proven implementations?
  - Are the used keys controlled by the customer? Where are they stored?
- Are you introducing any new policies/roles/users?
  - Have you used the least-privilege principle? How?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: backend/dataall/modules/dataset_sharing/db/share_object_repositories.py [hunk 1]
```diff
@@ -899,6 +899,36 @@ def find_all_share_items(session, share_uri, share_type, status=None):
             query = query.filter(ShareObjectItem.status.in_(status))
         return query.all()
 
+    @staticmethod
+    def find_all_other_share_items(
+        session, not_this_share_uri, item_uri, share_type, principal_type, principal_uri, item_status=None
+    ) -> List[ShareObjectItem]:
+        """
+        Find all shares from principal (principal_uri) to item (item_uri), that are not from specified share (not_this_share_uri)
+        """
```
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/db/share_object_repositories.py:908]: The 2 queries can be written as a single joined query

```
        shares = (
            session.query(ShareObject.shareUri)
            .join(
                ShareObjectItem,
                ShareObjectItem.shareUri == ShareObject.shareUri,
            )
            .filter(
                and_(
                    ShareObject.principalType == principal_type,
                    ShareObject.principalId == principal_uri,
                    ShareObject.shareUri != not_this_share_uri,
                    ShareObjectItem.itemUri == item_uri,
                    ShareObjectItem.itemType == share_type,
                )
            )
            .all()
        )
```

File: backend/dataall/modules/dataset_sharing/services/share_object_service.py [hunk 2]
```diff
@@ -533,13 +533,29 @@ def attach_dataset_table_read_permission(session, share):
             session, share.shareUri, ShareableType.Table.value, [ShareItemStatus.Share_Succeeded.value]
         )
         for table in share_table_items:
```
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/services/share_object_service.py:535]: Why the TODO?

**Review by dlpzx — APPROVED** [review]
Minor changes only. Overall looks great :)

File: backend/dataall/modules/dataset_sharing/services/share_object_service.py (hunk 2 above)
Comment (SofiaSazonova) [reply backend/dataall/modules/dataset_sharing/services/share_object_service.py:535]: I have some thoughts about this case and how to deal with it, but i think, we better discuss it first. Right now it's not necessary, so I decided to wait.

**Review by dlpzx — APPROVED** [review]
lgtm!

**Pull request merged**